jx/ui/Box.jxml

```diff
@@ -8,38 +8,61 @@
 			function calcLayout(delta, attr) {
 				// TODO: only calculate if something changed
 				var direction = delta.direction || attr.direction || 'vertical',
-					flow_dim = 'height', // flow dimension
+					flow_dim = 'height', // dimension in the flow direction
 					full_dim = 'width',  // other dimension (full size)
-					xy = 'y';            // positional axis
+					flow_axis = 'y',     // positional axis
+					full_axis = 'x',     // other axis (full size)
+					MIN_SIZE = 10,       // minimum element size
+					num_children = 0;    // number of children
 
 				if (direction == 'horizontal')
 					flow_dim = 'width',
 					full_dim = 'height',
-					xy = 'x';
+					flow_axis = 'x',
+					full_axis = 'y';
 
-				var my_size = delta[flow_dim] || attr[flow_dim] || 100,
-					full_size = delta[full_dim] || attr[full_dim] || 100;
+				var padding = int(delta['padding'] || attr['padding']) || 0,
+					border = int(delta['borderWidth'] || attr['borderWidth']) || 0,
+					my_size = (delta[flow_dim] || attr[flow_dim] || 100) - (padding + border) * 2,
+					full_size = (delta[full_dim] || attr[full_dim] || 100) - (padding + border) * 2;
 
 				delta.children = delta.children || {};
 
 				for (var k in attr.children)
 					delta.children[k] = delta.children[k] || {};
 
-				var count = 0;
+				var num_specified = 0, total_specified_size = 0;
 
-				for (k in delta.children)
-					count++;
+				for (k in delta.children) {
+					if (delta.children[k][flow_dim]) {
+						total_specified_size += int(delta.children[k][flow_dim]) || 0;
+						num_specified++;
+					}
+
+					num_children++;
+				}
 
-				var gap = delta.gap || attr.gap || 10,
-					size = ( my_size - gap * (count - 1) ) / count,
-					i = 0;
+				var gap = int(delta.gap || attr.gap) || 0,
+					avail_space = my_size - int(padding) * 2 - gap * (num_children - 1), // available space for all elements
```

Comment:
padding is already deducted from `my_size` above,  and is being deducted again here